<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 1 Bioconductor Package Submissions | Bioconductor Packages: Development, Maintenance, and Peer Review</title>
<meta name="author" content="Kevin Rue-Albrecht">
<meta name="author" content="Daniela Cassol">
<meta name="author" content="Johannes Rainer">
<meta name="author" content="Lori Shepherd">
<meta name="description" content="Introduction Types of Packages Package Naming Policy Author/Maintainer Expectations Submission Experiment data package Annotation package Workflow package Review Process Following Acceptance...">
<meta name="generator" content="bookdown 0.32 with bs4_book()">
<meta property="og:title" content="Chapter 1 Bioconductor Package Submissions | Bioconductor Packages: Development, Maintenance, and Peer Review">
<meta property="og:type" content="book">
<meta property="og:description" content="Introduction Types of Packages Package Naming Policy Author/Maintainer Expectations Submission Experiment data package Annotation package Workflow package Review Process Following Acceptance...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 1 Bioconductor Package Submissions | Bioconductor Packages: Development, Maintenance, and Peer Review">
<meta name="twitter:description" content="Introduction Types of Packages Package Naming Policy Author/Maintainer Expectations Submission Experiment data package Annotation package Workflow package Review Process Following Acceptance...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Bioconductor Packages: Development, Maintenance, and Peer Review</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">Welcome</a></li>
<li class="book-part">Package Submissions</li>
<li><a class="" href="submission-overview.html">Overview</a></li>
<li><a class="active" href="bioconductor-package-submissions.html"><span class="header-section-number">1</span> Bioconductor Package Submissions</a></li>
<li class="book-part">Package Development Guidelines</li>
<li><a class="" href="develop-overview.html">Overview</a></li>
<li><a class="" href="package-name.html"><span class="header-section-number">2</span> Package name</a></li>
<li><a class="" href="general.html"><span class="header-section-number">3</span> General Bioconductor Package Development</a></li>
<li><a class="" href="important-bioconductor-package-development-features.html"><span class="header-section-number">4</span> Important Bioconductor Package Development Features</a></li>
<li><a class="" href="readme.html"><span class="header-section-number">5</span> The README file</a></li>
<li><a class="" href="description.html"><span class="header-section-number">6</span> The DESCRIPTION file</a></li>
<li><a class="" href="namespace.html"><span class="header-section-number">7</span> The NAMESPACE file</a></li>
<li><a class="" href="news.html"><span class="header-section-number">8</span> The NEWS file</a></li>
<li><a class="" href="license.html"><span class="header-section-number">9</span> The LICENSE file</a></li>
<li><a class="" href="citation.html"><span class="header-section-number">10</span> The CITATION file</a></li>
<li><a class="" href="sysdep.html"><span class="header-section-number">11</span> The INSTALL file</a></li>
<li><a class="" href="docs.html"><span class="header-section-number">12</span> Documentation</a></li>
<li><a class="" href="data.html"><span class="header-section-number">13</span> Package data</a></li>
<li><a class="" href="tests.html"><span class="header-section-number">14</span> Unit tests</a></li>
<li><a class="" href="r-code.html"><span class="header-section-number">15</span> R code</a></li>
<li><a class="" href="other-than-Rcode.html"><span class="header-section-number">16</span> Fortran / C / C++ / Python / Third-Party Code</a></li>
<li><a class="" href="shiny.html"><span class="header-section-number">17</span> Shiny apps</a></li>
<li><a class="" href="non-software.html"><span class="header-section-number">18</span> Non-Software Packages</a></li>
<li><a class="" href="gitignore.html"><span class="header-section-number">19</span> The .gitignore file</a></li>
<li><a class="" href="conclusion.html"><span class="header-section-number">20</span> Conclusion</a></li>
<li class="book-part">Bioconductor Package Maintenance</li>
<li><a class="" href="package-maintenance.html">Overview</a></li>
<li><a class="" href="git-version-control.html"><span class="header-section-number">21</span> Git Version Control</a></li>
<li><a class="" href="versionnum.html"><span class="header-section-number">22</span> Version Numbering</a></li>
<li><a class="" href="troubleshooting-build-report.html"><span class="header-section-number">23</span> Troubleshooting Build Report</a></li>
<li><a class="" href="debugging-cc-code.html"><span class="header-section-number">24</span> Debugging C/C++ code</a></li>
<li><a class="" href="deprecation.html"><span class="header-section-number">25</span> Deprecation Guidelines</a></li>
<li><a class="" href="package-end-of-life-policy.html"><span class="header-section-number">26</span> Package End of Life Policy</a></li>
<li><a class="" href="branch-rename-faqs.html"><span class="header-section-number">27</span> Branch Rename FAQs</a></li>
<li class="book-part">Package Reviewer Resources</li>
<li><a class="" href="reviewer-resources-overview.html">Overview</a></li>
<li><a class="" href="review-expectation.html"><span class="header-section-number">28</span> Review Expectations</a></li>
<li><a class="" href="reviewtools.html"><span class="header-section-number">29</span> Reviewer Resources and Tools</a></li>
<li><a class="" href="review-volunteer-chapter.html"><span class="header-section-number">30</span> Volunteer to Review</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="use-devel.html"><span class="header-section-number">A</span> Using the ‘Devel’ Version of Bioconductor</a></li>
<li><a class="" href="long-tests.html"><span class="header-section-number">B</span> Long Tests</a></li>
<li><a class="" href="querying-web-resources.html"><span class="header-section-number">C</span> Querying Web Resources</a></li>
<li><a class="" href="c-fortran.html"><span class="header-section-number">D</span> C and Fortran code</a></li>
<li><a class="" href="cmavericks-best-practices.html"><span class="header-section-number">E</span> C++/Mavericks Best Practices</a></li>
<li><a class="" href="man-links.html"><span class="header-section-number">F</span> Debug: Links in Rd files</a></li>
<li><a class="" href="booknews.html"><span class="header-section-number">G</span> NEWS</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/Bioconductor/pkgrevdocs">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="bioconductor-package-submissions" class="section level1" number="1">
<h1>
<span class="header-section-number">1</span> Bioconductor Package Submissions<a class="anchor" aria-label="anchor" href="#bioconductor-package-submissions"><i class="fas fa-link"></i></a>
</h1>
<ul>
<li><a href="bioconductor-package-submissions.html#subintro">Introduction</a></li>
<li><a href="bioconductor-package-submissions.html#type">Types of Packages</a></li>
<li><a href="bioconductor-package-submissions.html#naming">Package Naming Policy</a></li>
<li><a href="bioconductor-package-submissions.html#author">Author/Maintainer Expectations</a></li>
<li><a href="bioconductor-package-submissions.html#submission">Submission</a></li>
<li><a href="bioconductor-package-submissions.html#experPackage">Experiment data package</a></li>
<li><a href="bioconductor-package-submissions.html#annPackage">Annotation package</a></li>
<li><a href="bioconductor-package-submissions.html#workPackage">Workflow package</a></li>
<li><a href="bioconductor-package-submissions.html#whattoexpect">Review Process</a></li>
<li><a href="bioconductor-package-submissions.html#afteraccept">Following Acceptance</a></li>
<li><a href="bioconductor-package-submissions.html#support">Additional Support</a></li>
</ul>
<div id="subintro" class="section level2" number="1.1">
<h2>
<span class="header-section-number">1.1</span> Introduction<a class="anchor" aria-label="anchor" href="#subintro"><i class="fas fa-link"></i></a>
</h2>
<p>To submit a package to <em>Bioconductor</em> the package should:</p>
<ul>
<li>Address areas of high-throughput genomic analysis, e.g., sequencing,
expression and other microarrays, flow cytometry, mass spectrometry,
image analysis; see <a href="https://bioconductor.org/packages/devel/BiocViews.html">biocViews</a>.</li>
<li>Interoperate with other <em>Bioconductor</em> packages by <em>re-using common data
structures</em> (see <a href="important-bioconductor-package-development-features.html#reusebioc">Common Bioconductor Methods and Classes</a>) and
existing infrastructure (e.g., <code>rtracklayer::import()</code> for input of common
genomic files).</li>
<li>Adopt software best practices that enable reproducible research and
use, such as full documentation and vignettes (including fully
evaluated code) as well as commitment to long-term user support
through the <em>Bioconductor</em> <a href="https://support.bioconductor.org/">support site</a>.</li>
<li>Not exist on CRAN. A package can only be submitted to one or the other.</li>
<li>Comply with <a href="develop-overview.html#develop-overview">Package Guidelines</a>.</li>
<li>Your package cannot depend on any package (or version of a package)
that is not (yet) available on CRAN or <em>Bioconductor</em>. The package should work
with whatever current version of the package is publically available.</li>
</ul>
</div>
<div id="type" class="section level2" number="1.2">
<h2>
<span class="header-section-number">1.2</span> Types of Packages<a class="anchor" aria-label="anchor" href="#type"><i class="fas fa-link"></i></a>
</h2>
<p><em>Bioconductor</em> packages are broadly defined by four main package types:
<a href="https://bioconductor.org/packages/release/bioc/"><strong>Software</strong></a>, <a href="https://bioconductor.org/packages/release/data/experiment/"><strong>Experiment Data</strong></a>,
<a href="https://bioconductor.org/packages/release/data/annotation/"><strong>Annotation</strong></a> and <a href="https://bioconductor.org/packages/release/workflows/"><strong>Workflow</strong></a>.</p>
<ul>
<li><p>Software Packages. Most packages contributed by users are software
packages. Software packages provide implementation of algorithms
(e.g. statistical analysis), access to resources (e.g. biomart, or NCBI) or
visualizations (e.g. volcano plots, pathways plots). Instructions for creating
Software packages can be found here: <a href="develop-overview.html#develop-overview">Package guidelines</a>.</p></li>
<li><p><a href="bioconductor-package-submissions.html#annPackage">Annotation packages</a> are database-like packages that provide
information linking identifiers (e.g., Entrez gene names or Affymetrix
probe ids) to other information (e.g., chromosomal location, Gene
Ontology category). It is also encouraged to utilize AnnotationHub for
storage and access to large raw data files and their conversion to
standard R formats. Instructions for adding data to AnnotationHub and
designing a annotation package to use AnnotationHub can be found here:
<a href="https://bioconductor.org/packages/devel/bioc/vignettes/HubPub/inst/doc/CreateAHubPackage.html">Create A Hub Package</a>.</p></li>
<li><p><a href="bioconductor-package-submissions.html#experPackage">Experiment data packages</a> provide data sets that are used,
often by software packages, to illustrate particular analyses. These packages
contain curated data from an experiment, teaching course or publication and
in most cases contain a single data set. It is also encouraged to utilize
ExperimentHub for storage and access to larger data files. ExperimentHub is
also particularly useful for hosting collections of related data sets.
Instructions for adding data to ExperimentHub and designing an experiment data
package to use ExperimentHub can be found here:
<a href="https://bioconductor.org/packages/devel/bioc/vignettes/HubPub/inst/doc/CreateAHubPackage.html">Create A Hub Package</a>.</p></li>
<li><p><a href="bioconductor-package-submissions.html#workPackage">Workflow packages</a> contain vignettes that describe a
bioinformatics workflow that involves multiple Bioconductor packages. These
vignettes are usually more extensive than the vignettes that accompany
software packages. These packages do not need man/ or R/ directories nor a
data/ directory as ideally workflows make use of existing data in a
Bioconductor package. See development section on <a href="non-software.html#workflow-packages">Workflows</a> for more
details.</p></li>
</ul>
<p>See <a href="develop-overview.html#develop-overview">Package Guidelines</a> for details on package format and syntax.</p>
</div>
<div id="naming" class="section level2" number="1.3">
<h2>
<span class="header-section-number">1.3</span> Package Naming Policy<a class="anchor" aria-label="anchor" href="#naming"><i class="fas fa-link"></i></a>
</h2>
<p>Package naming:</p>
<ol style="list-style-type: lower-roman">
<li><p>Ownership of package name. Bioconductor follows <a href="https://cran.r-project.org/web/packages/policies.html">CRAN’s
policy</a> in requiring that contributors give the right to use the package name
to Bioconductor at time of submission, so that the Bioconductor team can orphan
the package and allow another maintainer to take it over in the event that the
package contributor discontinues package maintenance. See Bioconductor’s package
<a href="package-end-of-life-policy.html#package-end-of-life-policy">end-of-life policy</a> for more details.</p></li>
<li><p>Uniqueness of package name. Packages should be named in a way that does not
conflict (irrespective of case) with any current or past BIOCONDUCTOR package,
nor any current or past CRAN package.</p></li>
</ol>
<p>See <a href="package-name.html#package-name">Package naming guidelines</a> for more guidelines.</p>
</div>
<div id="author" class="section level2" number="1.4">
<h2>
<span class="header-section-number">1.4</span> Author / Maintainer Expectations<a class="anchor" aria-label="anchor" href="#author"><i class="fas fa-link"></i></a>
</h2>
<p>Acceptance of packages into <em>Bioconductor</em> brings with it ongoing
package maintenance responsibilities. Package authors are expected to:</p>
<ul>
<li>
<p>Follow Bioconductor guidelines</p>
<p>These include standard <a href="develop-overview.html#develop-overview">guidelines</a>, <a href="versionnum.html#versionnum">version numbering</a>,
<a href="r-code.html#r-code">coding style</a>, <a href="general.html#correctness-space-and-time">code performance
requirements</a>, <a href="general.html#memory">memory usage</a>, <a href="important-bioconductor-package-development-features.html#reusebioc">using
existing data classes</a>, and the other requirements described
below.</p>
</li>
<li>
<p>Follow the release cycle of Bioconductor</p>
<p>There are two releases each year, around April and October. The
<a href="https://bioconductor.org/developers/release-schedule/">release schedule</a> will indicate the timetables and
deadlines for each release. A release cycle typically produces two
versions of packages, ‘devel’ and ‘release’. It is important to be familiar
with these branch concepts. Once your package has been accepted, it will
initially be in the ‘devel’ branch. The current devel branch becomes the
next release. Most users are expected to use the release branch, so they will
not immediately have access to your package until the next release.
Bug fixes can be fixed in both branches, while new features should only
be added to the ‘devel’ branch.</p>
</li>
<li>
<p>Maintain the package using version control</p>
<p>Realize that <em>Bioconductor</em>, unlike CRAN, maintains all package source code
under git version control. This means that you make changes to your
package using <a href="git-version-control.html#git-version-control">git</a>. If your package is accepted, you will receive
instructions with typical git operations (see the after
<a href="bioconductor-package-submissions.html#afteraccept">acceptance section</a>). Package maintenance through software
release cycles, including prompt updates to software and documentation,
is needed due to possible underlying changes in R and/or other package
dependencies.</p>
</li>
<li>
<p>Check Build Reports and Fix Issues Promptly</p>
<p><em>Bioconductor</em> has weekly to daily <a href="https://bioconductor.org/checkResults/">build reports</a> for all package types. It is
the maintainers responsibility to check the build reports and respond to any
ERROR/Warning/Notes produced by the install, build, or check process of a
package. The maintainer will receive automatic build notifications when the
package starts to fail on linux from <code>BBS-noreply@bioconductor.org</code>;
maintainers should check that emails from this email address can be delivered.</p>
</li>
<li>
<p>Subscribe to the <a href="https://stat.ethz.ch/mailman/listinfo/bioc-devel">bioc-devel</a> mailing list</p>
<p>The Bioconductor team communicates with developers through this list.
It is also a good channel to communicate changes to other developers.
Addressing Bioconductor team requests in a timely manner guarantees that
your package remains available through Bioconductor.</p>
</li>
<li>
<p><a href="https://support.bioconductor.org/accounts/signup/">Register</a> on and use the <a href="https://support.bioconductor.org/">support site</a></p>
<p>The support site is the official support channel for users. Users and even
developers may ask questions regarding your package on this platform.
Be sure to include all the packages that you maintain in the “Watched Tags”
section of your support site profile. This will notify you of any questions
posted regarding your package(s).
It is important to promptly respond to bug reports and questions either on
the <a href="https://support.bioconductor.org/"><em>Bioconductor</em> support site</a> post or directly to developers.
Some maintainers prefer to indicate a <code>BugReports:</code> field in their package’s
DESCRIPTION file. This field indicates a particular web page for submitting
bug reports and questions.</p>
</li>
</ul>
</div>
<div id="submission" class="section level2" number="1.5">
<h2>
<span class="header-section-number">1.5</span> Submission<a class="anchor" aria-label="anchor" href="#submission"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><p>Read and follow the full <a href="develop-overview.html#develop-overview">Contributor Guidelines</a> section.</p></li>
<li><p>Submit by opening a new issue in the <em>Bioconductor</em>
<a href="https://github.com/Bioconductor/Contributions/issues/new">Contributions</a> repository, following the <a href="https://github.com/Bioconductor/Contributions">guidelines</a> of
the <code>README.md</code> file.
Assuming that your package is in a <a href="https://help.github.com/articles/create-a-repo/">GitHub Repository</a> and
under the default branch, add the link to your repository to
the issue you are opening. You cannot specify any alternative branches; the
default branch is utilized. The default branch must contain only package
code. Any files or directories for other applications (Github Actions,
devtool, etc) should be in a different branch. If you are submitting two
highly related packages or circular dependent packages please see
<a href="https://github.com/Bioconductor/Contributions/blob/master/CONTRIBUTING.md#submitting-related-packages">here</a>. The lighter dependent or package that can be installed without
a dependency should be submitted first; this is generally the associated data
package to a software package.</p></li>
</ul>
</div>
<div id="experPackage" class="section level2" number="1.6">
<h2>
<span class="header-section-number">1.6</span> Experiment Data Packages<a class="anchor" aria-label="anchor" href="#experPackage"><i class="fas fa-link"></i></a>
</h2>
<p>Experimental data packages contain data specific to a particular
analysis or experiment. They often accompany a software package for
use in the examples and vignettes and in general are not updated
regularly. If you need a general subset of data for workflows or
examples first check the AnnotationHub resource for available files
(e.g., BAM, FASTA, BigWig, etc.) or ExperimentHub for available processed
example data set already included in <em>Bioconductor</em>. If no current files or data
sets are appropriate consider an associated Experiment Data Package that
utilizes <em><a href="https://bioconductor.org/packages/3.17/ExperimentHub">ExperimentHub</a></em>.</p>
<p>If you have an associated data package for your software package,
please do <em>NOT</em> create a separate issue in the our tracker repository
for that. Instead, please add the data package repository to the same
issue as the software package. The process for doing this is
documented <a href="https://github.com/Bioconductor/Contributions/blob/master/CONTRIBUTING.md#submitting-related-packages">here</a>. Generally the data package should be submitted first.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.17/HubPub">HubPub</a></em> package is helpful for creating a template
for a hub package. The vignette <a href="https://bioconductor.org/packages/devel/bioc/vignettes/HubPub/inst/doc/CreateAHubPackage.html">Create A Hub Package</a> provides full details.</p>
</div>
<div id="annPackage" class="section level2" number="1.7">
<h2>
<span class="header-section-number">1.7</span> Annotation Packages<a class="anchor" aria-label="anchor" href="#annPackage"><i class="fas fa-link"></i></a>
</h2>
<p>Annotation packages contain lightly or non-curated data from a public
source and are updated with each <em>Bioconductor</em> release (every 6
months). They are a source of general annotation for one or many
organisms and are not specific to a particular experiment. When
possible, they should support the <code>select()</code> interface from
AnnotationDbi.</p>
<p>Annotation packages should <em>NOT</em> be posted to the tracker repository.
Instead send an email to <a href="mailto:packages@bioconductor.org" class="email">packages@bioconductor.org</a> with a
description of the proposed annotation package and futher instructions
of where to send the package will be provided. Whenever possible Annotation
Packages should use the <em><a href="https://bioconductor.org/packages/3.17/AnnotationHub">AnnotationHub</a></em> for managing
files.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.17/HubPub">HubPub</a></em> package is helpful for creating a template
for a hub package. The vignette <a href="https://bioconductor.org/packages/devel/bioc/vignettes/HubPub/inst/doc/CreateAHubPackage.html">Create A Hub Package</a> provides full details.</p>
<p><a name="workPackage"></a></p>
</div>
<div id="workPackage" class="section level2" number="1.8">
<h2>
<span class="header-section-number">1.8</span> Workflow Packages<a class="anchor" aria-label="anchor" href="#workPackage"><i class="fas fa-link"></i></a>
</h2>
<p>Workflow packages contain vignettes that describe a bioinformatics workflow that
involves multiple Bioconductor packages. These vignettes are usually more
extensive than the vignettes that accompany software packages. These packages do
not need man/ or R/ directories nor a data/ directory as ideally workflows make
use of existing data in a Bioconductor package. See development section on
<a href="non-software.html#workflow-packages">Workflows</a> for more details</p>
<p><a name="whattoexpect"></a></p>
</div>
<div id="whattoexpect" class="section level2" number="1.9">
<h2>
<span class="header-section-number">1.9</span> Review Process<a class="anchor" aria-label="anchor" href="#whattoexpect"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><p>A new package is initially labeled as <code>1. awaiting moderation</code>.
A <em>Bioconductor</em> team member will take a very brief look at your
package, to ensure that it is intended for <em>Bioconductor</em>. Appropriate
packages will be re-labelled <code>2. review in progress</code> and a reviewer will be
automatically assigned. Your assigned reviewer will address your concerns and
help you through the review process. The entire review process typically
takes between 2 and 6 weeks. If there is no response after 3 to 4 weeks,
package reviewers may close the issue until further updates, changes, and/or
commentary are received.</p></li>
<li><p>If in the pre-review process there are identified larger issues with the
package, a label <code>3e. pending pre-review changes</code> or a more specific flag of
package issue will be assigned. Please address any pre-review identified
issues and comment back for the package admininstrators to re-evaluate.</p></li>
<li><p>Please be courteous with your package reviewers and always follow the
<em>Bioconductor</em> <a href="https://bioconductor.github.io/bioc_coc_multilingual/">Code Of Conduct</a> with correspondence. Please allow 2-3 weeks
for reviewers to assess your package.</p></li>
<li><p>The package will be submitted to the <em>Bioconductor</em> build
system (BBS). The system will check out your package from GitHub and move it
to our git.bioconductor.org git server. Please familiarize yourself with
<a href="git-version-control.html#git-version-control">git</a> as the git.bioconductor.org is the versions the (BBS) will
always use. It will then run <code>R CMD build</code> to create a ‘tarball’ of your source code,
vignettes, and man pages. It will run <code>R CMD check</code> on the tarball,
to ensure that the package conforms to standard <em>R</em> programming best
practices. <em>Bioconductor</em> has chosen to utilize a custom <code>R CMD check</code>
environment; See <a href="general.html#r-cmd-check-environment">R CMD check environment</a> for more details. Finally, the
build system will run <code>BiocCheckGitClone()</code> and <code>BiocCheck('new-package'=TRUE)</code> to ensure that
the package conforms to <em>Bioconductor</em> <em><a href="https://bioconductor.org/packages/3.17/BiocCheck">BiocCheck</a></em>
standards. The system will perform these steps using the <a href="use-devel.html#use-devel">‘devel’
version</a> of <em>Bioconductor</em>, on three platforms
(Linux, Mac OS X, and Windows). After these steps are complete, a link to a
build report will be appended to the new package issue. Avoid surprises by
running these checks on your own computer, under the ‘devel’ version
of <em>Bioconductor</em>, before submitting your package.</p></li>
<li><p>If the build report indicates problems, modify your package and
commit changes to the git.bioconductor.org version of your package as
described in the <a href="git-version-control.html#new-package-workflow">new package git workflow</a>. If
there are problems that you do not understand, seek help on the
<a href="https://stat.ethz.ch/mailman/listinfo/bioc-devel">bioc-devel</a> mailing list.</p></li>
<li><p>To trigger a new build, include a version bump in your commit, e.g.,
from <code>Version: 0.99.0</code> to <code>Version: 0.99.1</code>. Pre-release versions utilize the
<code>0.99.z</code> format. When accepted and released, your package’s version number
will be automatically incremented to 1.0.0.</p></li>
<li><p>Once your package builds and checks without errors or (avoidable)
warnings, a <em>Bioconductor</em> team member will provide a technical
review of your package. Other <em>Bioconductor</em> developers and users
with domain expertise are encouraged to provide additional community
commentary. Reviewers will add comments to the issue you created.</p></li>
<li><p>Respond to the issues raised by the reviewers. You <em>must</em> respond to
the primary reviewer, and are strongly encouraged to consider
community commentary. Typically your response will involve code
modifications; commit these to the default branch of git.bioconductororg. When
you have addressed all concerns, add a comment to the issue created in step 2
to explain your response.</p></li>
<li><p>The reviewer will assess your responses, perhaps suggesting further
modifications or clarification. The reviewer will then accept your
package for inclusion in <em>Bioconductor</em>, or decline it. The label
<code>2. review in progress</code> will be replaced by <code>3a. accepted</code> or
<code>3b. declined</code>.</p></li>
<li><p>If your package is accepted, it will be added to <em>Bioconductor</em>’s
nightly ‘devel’ builds. All packages in the ‘devel’ branch of the repository
are ‘released’ to the user community once every six months, in
approximately April and October.</p></li>
<li><p>Once the review process is complete, the issue you created will be
closed. All updates to your package will be through the
<a href="git-version-control.html#git-version-control"><em>Bioconductor</em> Git Server</a>.</p></li>
</ul>
</div>
<div id="afteraccept" class="section level2" number="1.10">
<h2>
<span class="header-section-number">1.10</span> Following Acceptance<a class="anchor" aria-label="anchor" href="#afteraccept"><i class="fas fa-link"></i></a>
</h2>
<p>Following acceptance of a package:</p>
<ul>
<li>Packages accepted on the tracker repository are added to the ‘devel’
branch of the <em>Bioconductor</em> GIT repository, with the current version
number of the accepted package.</li>
<li>Packages are then built by the <em>Bioconductor</em> nightly build
process. On-demand builds of accepted packages do not occur. Please see the
<a href="https://bioconductor.org/checkResults/">build reports</a> for how often the different package types are built. The
changes pushed to a ‘devel’ version of a package can take up to 24-28 hours to
appear. See <a href="troubleshooting-build-report.html#propagate">build timings</a>. If the build is successful, the package has its own
‘landing page’ created, and the package is made available to users
of the ‘devel’ branch of <em>Bioconductor</em> via <code><a href="https://rdrr.io/pkg/BiocManager/man/install.html">BiocManager::install()</a></code>.</li>
<li>Changes to their package (if any), should be done to version in the
<a href="git-version-control.html#git-version-control"><em>Bioconductor</em> git server</a>.</li>
<li>Developers should bump the <code>z</code> portion of their version number every
time they commit changes to their package, following the
<a href="versionnum.html#versionnum">Version numbering</a> guidelines. If developers don’t
bump the version, the changes made to their package <em>do not propagate</em>
to the <em>Bioconductor</em> web site and package repository.</li>
</ul>
</div>
<div id="support" class="section level2" number="1.11">
<h2>
<span class="header-section-number">1.11</span> Additional Support<a class="anchor" aria-label="anchor" href="#support"><i class="fas fa-link"></i></a>
</h2>
<p>We are eager to enhance the quality and interoperability of
<em>Bioconductor</em> software and will provide additional support when
requested by package developers. Example areas of assistance include
use of appropriate S4 structures, specific guidance on efficient
implementation, guidance on code structure, and critical assessment of
package documentation and structure. Use the
<a href="https://stat.ethz.ch/mailman/listinfo/bioc-devel">bioc-devel</a> mailing list or email
<a href="mailto:maintainer@bioconductor.org" class="email">maintainer@bioconductor.org</a> to obtain additional support.</p>
<ul>
<li>Support Email: <a href="mailto:maintainer@bioconductor.org" class="email">maintainer@bioconductor.org</a>
</li>
</ul>
</div>
</div>



<div class="chapter-nav">
<div class="prev"><a href="submission-overview.html">Overview</a></div>
<div class="next"><a href="develop-overview.html">Overview</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#bioconductor-package-submissions"><span class="header-section-number">1</span> Bioconductor Package Submissions</a></li>
<li><a class="nav-link" href="#subintro"><span class="header-section-number">1.1</span> Introduction</a></li>
<li><a class="nav-link" href="#type"><span class="header-section-number">1.2</span> Types of Packages</a></li>
<li><a class="nav-link" href="#naming"><span class="header-section-number">1.3</span> Package Naming Policy</a></li>
<li><a class="nav-link" href="#author"><span class="header-section-number">1.4</span> Author / Maintainer Expectations</a></li>
<li><a class="nav-link" href="#submission"><span class="header-section-number">1.5</span> Submission</a></li>
<li><a class="nav-link" href="#experPackage"><span class="header-section-number">1.6</span> Experiment Data Packages</a></li>
<li><a class="nav-link" href="#annPackage"><span class="header-section-number">1.7</span> Annotation Packages</a></li>
<li><a class="nav-link" href="#workPackage"><span class="header-section-number">1.8</span> Workflow Packages</a></li>
<li><a class="nav-link" href="#whattoexpect"><span class="header-section-number">1.9</span> Review Process</a></li>
<li><a class="nav-link" href="#afteraccept"><span class="header-section-number">1.10</span> Following Acceptance</a></li>
<li><a class="nav-link" href="#support"><span class="header-section-number">1.11</span> Additional Support</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/Bioconductor/pkgrevdocs/blob/master/package-submission.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/Bioconductor/pkgrevdocs/edit/master/package-submission.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>
</div>
  

  

</div>
 <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Bioconductor Packages: Development, Maintenance, and Peer Review</strong>" was written by Kevin Rue-Albrecht, Daniela Cassol, Johannes Rainer, Lori Shepherd. It was last built on 2023-01-19.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
