# Important _Bioconductor_ Package Development Features {#important-bioconductor-package-development-features}

## biocViews {#biocviews}

Packages added to the _Bioconductor_ Project require a `biocViews:`
field in their DESCRIPTION file.  The field name "biocViews" is
case-sensitive and must begin with a lower-case 'b'.

biocViews terms are "keywords" used to describe a given package. They
are broadly divided into four categories, representing the type of
packages present in the _Bioconductor_ Project -

1. Software
2. Annotation Data
3. Experiment Data
4. Workflow

biocViews are available for the [release][] and [devel][] branches of
_Bioconductor_. The devel branch has a check box under the tree
structure which, when checked, displays biocViews that are defined but
not used by any package, in addition to biocViews that are in use. See also
[description section](#description-biocviews)

### Motivation {#biocviews-motivation}

One can use biocViews for two broad purposes.

1. A researcher might want to identify all packages in the
   _Bioconductor_ Project which are related to a specific purpose.
   For example, one may want to look for all packages related to "Copy
   Number Variants".

2. During development, a package contributor can "tag" their package
   with biocViews so that when someone looking for packages (like in
   scenario 1) can easily find their package.

### biocViews during new package development {#biocviews-pkg-devel}

Visit the [devel][] branch of biocViews when you are in the process of
adding biocViews to your new package. Identify as many terms as
appropriate from the hierarchy. Prefer 'leaf' terms at the end of the
hierarchy, over more inclusive terms. Remember to check the box
displaying all available terms.

Please Note:

1. Your package will belong to only one part of _Bioconductor_ Project
   (Software, Annotation Data, Experiment Data, Workflow), so choose only
   biocViews from that category.

2. biocViews listed in your package must match exactly (e.g.,
   spelling, capitalization) the terms in the biocViews hierarchy.

When you submit your new package for review , your package is checked
and built by the _Bioconductor_ Project.  We check the following for
biocViews:

1. Package contributor has added biocViews.

2. biocViews are valid.

3. Package contributor has added biocViews from only one of the categories.

If you receive a "RECOMMENDED" direction for any of these biocViews
after you have submitted your package, you can try correcting them on
your own following the directions given here or ask your package
reviewer for more information.

If a developer thinks a biocViews term should be added to the current acceptable
list, please email bioc-devel@r-project.org requesting the new biocView, under
which hierarchy the term should be placed, and the justification for the new
term.

[release]: http://bioconductor.org/packages/release/BiocViews.html
[devel]: http://bioconductor.org/packages/devel/BiocViews.html


## Common Bioconductor Methods and Classes {#reusebioc}

We strongly recommend reusing existing methods for importing data, and
reusing established classes for representing data. Here are some
suggestions for importing different file types and commonly used
_Bioconductor_ classes. For more classes and functionality also try
searching in [BiocViews](/packages) for your data type.

### Importing {#commonimport}

+ GTF, GFF, BED, BigWig, etc., -- [rtracklayer][]`::import()`
+ VCF -- [VariantAnnotation][]`::readVcf()`
+ SAM / BAM -- [Rsamtools][]`::scanBam()`,
  [GenomicAlignments][]`::readGAlignment*()`
+ FASTA -- [Biostrings][]`::readDNAStringSet()`
+ FASTQ -- [ShortRead][]`::readFastq()`
+ MS data (XML-based and mgf formats) -- [Spectra][]`::Spectra()`, 
  [MSnbase][]`::readMSData()`, [Spectra][]`::Spectra(source = MsBackendMgf::MsBackendMgf())`, 
  [MSnbase][]`::readMgfData()`

### Common Classes {#commonclass}

+ Rectangular feature x sample data --
  [SummarizedExperiment][]`::SummarizedExperiment()` (RNAseq count
  matrix, microarray, ...)
+ Genomic coordinates -- [GenomicRanges][]`::GRanges()` (1-based,
  closed interval)
+ Genomic coordinates from multiple samples --
  [GenomicRanges][]`::GRangesList()`
+ Ragged genomic coordinates -- [RaggedExperiment][]`::RaggedExperiment()`
+ DNA / RNA / AA sequences -- [Biostrings][]`::*StringSet()`
+ Gene sets -- [BiocSet][]`::BiocSet()`,
  [GSEABase][]`::GeneSet()`,
  [GSEABase][]`::GeneSetCollection()`
+ Multi-omics data --
  [MultiAssayExperiment][]`::MultiAssayExperiment()`
+ Single cell data --
  [SingleCellExperiment][]`::SingleCellExperiment()`
+ Mass spec data -- [Spectra][]`::Spectra()`, [MSnbase][]`::MSnExp()`


In general, a package will not be accepted if it does not show interoperability
with the current _Bioconductor_ ecosystem.

[rtracklayer]: https://bioconductor.org/packages/rtracklayer
[Biostrings]: https://bioconductor.org/packages/Biostrings
[Rsamtools]: https://bioconductor.org/packages/Rsamtools
[GenomicAlignments]: https://bioconductor.org/packages/GenomicAlignments
[VariantAnnotation]: https://bioconductor.org/packages/VariantAnnotation
[ShortRead]: https://bioconductor.org/packages/ShortRead
[MSnbase]: https://bioconductor.org/packages/MSnbase
[SummarizedExperiment]: https://bioconductor.org/packages/SummarizedExperiment
[GenomicRanges]: https://bioconductor.org/packages/GenomicRanges
[RaggedExperiment]: https://bioconductor.org/packages/RaggedExperiment
[BiocSet]: https://bioconductor.org/packages/BiocSet
[GSEABase]: https://bioconductor.org/packages/GSEABase
[MultiAssayExperiment]: https://bioconductor.org/packages/MultiAssayExperiment
[SingleCellExperiment]: https://bioconductor.org/packages/SingleCellExperiment
[Spectra]: https://bioconductor.org/packages/Spectra
[MsBackendMgf]: https://bioconductor.org/packages/MsBackendMgf

