<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 24 Non-Software Packages | Bioconductor Packages: Development, Maintenance, and Peer Review</title>
<meta name="author" content="Kevin Rue-Albrecht">
<meta name="author" content="Daniela Cassol">
<meta name="author" content="Johannes Rainer">
<meta name="author" content="Lori Shepherd">
<meta name="description" content="Most packages contributed by users are software packages. However, there are instances where other package types are submitted. The following sections will go into specifics we look for in each of...">
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Chapter 24 Non-Software Packages | Bioconductor Packages: Development, Maintenance, and Peer Review">
<meta property="og:type" content="book">
<meta property="og:description" content="Most packages contributed by users are software packages. However, there are instances where other package types are submitted. The following sections will go into specifics we look for in each of...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 24 Non-Software Packages | Bioconductor Packages: Development, Maintenance, and Peer Review">
<meta name="twitter:description" content="Most packages contributed by users are software packages. However, there are instances where other package types are submitted. The following sections will go into specifics we look for in each of...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS -->
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Bioconductor Packages: Development, Maintenance, and Peer Review</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">Welcome</a></li>
<li class="book-part">Package Development Guidelines</li>
<li><a class="" href="develop-overview.html">Overview</a></li>
<li><a class="" href="package-name.html"><span class="header-section-number">1</span> Package name</a></li>
<li><a class="" href="general.html"><span class="header-section-number">2</span> General Package Development</a></li>
<li><a class="" href="use-devel.html"><span class="header-section-number">3</span> Using the ‘Devel’ Version of Bioconductor</a></li>
<li><a class="" href="description.html"><span class="header-section-number">4</span> The DESCRIPTION file</a></li>
<li><a class="" href="biocviews-1.html"><span class="header-section-number">5</span> biocViews</a></li>
<li><a class="" href="version-numbering.html"><span class="header-section-number">6</span> Version Numbering</a></li>
<li><a class="" href="namespace.html"><span class="header-section-number">7</span> The NAMESPACE file</a></li>
<li><a class="" href="news.html"><span class="header-section-number">8</span> The NEWS file</a></li>
<li><a class="" href="citation.html"><span class="header-section-number">9</span> The CITATION file</a></li>
<li><a class="" href="data.html"><span class="header-section-number">10</span> Package data</a></li>
<li><a class="" href="docs.html"><span class="header-section-number">11</span> Documentation</a></li>
<li><a class="" href="tests.html"><span class="header-section-number">12</span> Unit tests</a></li>
<li><a class="" href="long-tests.html"><span class="header-section-number">13</span> Long Tests</a></li>
<li><a class="" href="troubleshooting-build-report.html"><span class="header-section-number">14</span> Troubleshooting Build Report</a></li>
<li><a class="" href="r-code.html"><span class="header-section-number">15</span> R code</a></li>
<li><a class="" href="common-bioconductor-methods-and-classes.html"><span class="header-section-number">16</span> Common Bioconductor Methods and Classes</a></li>
<li><a class="" href="robust-and-efficient-code.html"><span class="header-section-number">17</span> Robust and Efficient Code</a></li>
<li><a class="" href="c-fortran.html"><span class="header-section-number">18</span> C and Fortran code</a></li>
<li><a class="" href="cmavericks-best-practices.html"><span class="header-section-number">19</span> C++/Mavericks Best Practices</a></li>
<li><a class="" href="debugging-cc-code.html"><span class="header-section-number">20</span> Debugging C/C++ code</a></li>
<li><a class="" href="querying-web-resources.html"><span class="header-section-number">21</span> Querying Web Resources</a></li>
<li><a class="" href="third-party-code.html"><span class="header-section-number">22</span> Third-party code</a></li>
<li><a class="" href="shiny.html"><span class="header-section-number">23</span> Shiny apps</a></li>
<li><a class="active" href="non-software.html"><span class="header-section-number">24</span> Non-Software Packages</a></li>
<li><a class="" href="package-end-of-life-policy.html"><span class="header-section-number">25</span> Package End of Life Policy</a></li>
<li><a class="" href="gitignore.html"><span class="header-section-number">26</span> The .gitignore file</a></li>
<li><a class="" href="git-version-control.html"><span class="header-section-number">27</span> Git Version Control</a></li>
<li><a class="" href="conclusion.html"><span class="header-section-number">28</span> Conclusion</a></li>
<li><a class="" href="references-1.html">References</a></li>
<li class="book-part">Package Submissions</li>
<li><a class="" href="software-package-submissions.html"><span class="header-section-number">29</span> Software Package Submissions</a></li>
<li class="book-part">Package Reviewer Resources</li>
<li><a class="" href="reviewer-resources-overview.html">Overview</a></li>
<li><a class="" href="review-expectation.html"><span class="header-section-number">30</span> Review Expectations</a></li>
<li><a class="" href="reviewtools.html"><span class="header-section-number">31</span> Reviewer Resources and Tools</a></li>
<li><a class="" href="review-volunteer-chapter.html"><span class="header-section-number">32</span> Volunteer to Review</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="booknews.html"><span class="header-section-number">A</span> NEWS</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/Bioconductor/pkgrevdocs">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="non-software" class="section level1" number="24">
<h1>
<span class="header-section-number">24</span> Non-Software Packages<a class="anchor" aria-label="anchor" href="#non-software"><i class="fas fa-link"></i></a>
</h1>
<p>Most packages contributed by users are software packages. However, there are
instances where other package types are submitted. The following sections will
go into specifics we look for in each of the non-software type packages.</p>
<div id="annotation-and-experiment-data-packages" class="section level2" number="24.1">
<h2>
<span class="header-section-number">24.1</span> Annotation and Experiment data packages<a class="anchor" aria-label="anchor" href="#annotation-and-experiment-data-packages"><i class="fas fa-link"></i></a>
</h2>
<p>Annotation packages are database-like packages that provide information linking
identifies (e.g., Entrez gene names or Affymetrix probe ids) to other
information (e.g., chromosomal location, Gene Ontology category).</p>
<p>Experiment data packages provide data sets that are used, often by software
packages, to illustrate particular analyses. These packages contain curated
data from an experiment, teaching course or publication and in most cases
contain a single data set.</p>
<p>We look for similar requirements as software packages, but most importantly is
proper documentation for the data included within the package. Traditional
Annotation and Experiment packages are not ideal; AnnotationHub and
ExperimentHub interfaces and packages are desirable.</p>
</div>
<div id="annotationexperiment-hub-packages" class="section level2" number="24.2">
<h2>
<span class="header-section-number">24.2</span> Annotation/Experiment Hub packages<a class="anchor" aria-label="anchor" href="#annotationexperiment-hub-packages"><i class="fas fa-link"></i></a>
</h2>
<p>These light weight packages are related to resources added to AnnotationHub or
ExperimentHub. The package should minimally contain the resource metadata, man
pages describing the resources, and a vignette. It may also contain supporting
R function the author wants to provide. These packages are similar to the
above Annotation or Experiment data packages except the data is stored in AWS S3
buckets or publicly accessibly sites instead of in the package itself.</p>
<p>There is more information about creating a hub packages as well as the contents
of on in the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/HubPub/inst/doc/CreateAHubPackage.html">Creating a Hub Package</a> vignette within the <code>HubPub</code> Bioconductor
package.</p>
</div>
<div id="workflow-packages" class="section level2" number="24.3">
<h2>
<span class="header-section-number">24.3</span> Workflow packages<a class="anchor" aria-label="anchor" href="#workflow-packages"><i class="fas fa-link"></i></a>
</h2>
<p>Workflow packages contain vignettes that describe a bioinformatics workflow
that involves multiple Bioconductor packages. These vignettes are usually more
extensive than the vignettes that accompany software packages. These packages do
not need <code>man/</code> or <code>R/</code> directories nor a <code>data/</code> directory as ideally
workflows make use of existing data in a Bioconductor package.</p>
<p><a href="http://bioconductor.org/packages/devel/BiocViews.html#___Workflow">Existing Workflows</a></p>
<p>Workflow vignettes may deal with larger data sets and/or be more computationally intensive
than typical Bioconductor package vignettes. For this reason, the automated builder that
produces these vignettes does not have a time limit (in contrast to the Bioconductor package
building system which will time out if package building takes too long). It is
expected the majority of vignette code chunks are evaluated.</p>
<div id="how-do-i-write-and-submit-a-workflow-vignette" class="section level3" number="24.3.1">
<h3>
<span class="header-section-number">24.3.1</span> How do I write and submit a workflow vignette?<a class="anchor" aria-label="anchor" href="#how-do-i-write-and-submit-a-workflow-vignette"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li><p>Write a package with the same name as the workflow. The workflow vignette
written in Markdown, using the <a href="http://rmarkdown.rstudio.com/">rmarkdown</a>
package should be included in the vignette directory. You may include more than
one vignette but please use useful identifying names.</p></li>
<li><p>The package does not need man/ or R/ directories nor a data/ directory as
ideally workflows make use of existing data in a Bioconductor repository or on
the web; the workflow package itself should not contain large data files.</p></li>
<li><p>In the DESCRIPTION file, include the line “BiocType: Workflow.” Please also
include a detailed Description field in the DESCRIPTION file. The DESCRIPTION
file should contain biocViews which should be from the <a href="http://bioconductor.org/packages/devel/BiocViews.html#___Workflow">Workflow
branch</a>. If
you think a new term is relevant please reach out to
<a href="mailto:lori.shepherd@roswellpark.org" class="email">lori.shepherd@roswellpark.org</a>.</p></li>
<li><p>Submit the package to the <a href="https://github.com/Bioconductor/Contributions">GitHub submission
tracker</a> for a formal
review. Please also indicate in the tracker issue that this package is a
workflow.</p></li>
<li><p>Workflows are git version controlled. Once the package is accepted it will be
added to our git repository at <a href="mailto:git@git.bioconductor.org" class="email">git@git.bioconductor.org</a> and instructions will
be sent for gaining access for maintainence.</p></li>
</ul>
</div>
<div id="consistent-formatting" class="section level3" number="24.3.2">
<h3>
<span class="header-section-number">24.3.2</span> Consistent formatting<a class="anchor" aria-label="anchor" href="#consistent-formatting"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>
<p>In an effort to standardize the workflow vignette format, it is strongly
encouraged to use either BiocStyle for formatting or utilize
<a href="http://bioconductor.org/packages/BiocWorkflowTools/">BiocWorkflowTools</a>.
The following header shows how to use BiocStyle in the vignette:</p>
<pre><code>output:
    BiocStyle::html_document</code></pre>
</li>
<li>
<p>The following should also be include</p>
<pre><code>- author affiliations
- a date representing when the workflow vignette has been modified</code></pre>
</li>
<li>
<p>The first section should have some versioning information. The R version,
Bioconductor version, and package version should be visible.
The following is an example of how this could be achieved:</p>
<pre><code>&lt;p&gt;
**R version**: R Under development (unstable) (2022-01-29 r81593)
&lt;br /&gt;
**Bioconductor version**: 3.15
&lt;br /&gt;
**Package**: 1.19.0
&lt;\p&gt;</code></pre>
</li>
<li><p>An example start to a workflow vignette:</p></li>
</ul>
<p>The following is taken as an example header from the variants workflow package:</p>
<pre>
    <code>
– – –
title: Annotating Genomic Variants
author: 
–name: Valerie Obenchain
  affiliation: Fred Hutchinson Cancer Research Center, 1100 Fairview Ave. N., P.O. Box 19024, Seattle, WA, USA 98109–1024
date: 11 April 2018
vignette: &gt;
  %\VignetteIndexEntry{Annotating Genomic Variants}
  %\VignetteEngine{knitr::rmarkdown}
output: 
    BiocStyle::html_document
– – –


# Version Info
```{r, echo=FALSE, results="hide", warning=FALSE}
suppressPackageStartupMessages({library('variants')})
```
&lt;p&gt;
**R version**: `r R.version.string`
&lt;br /&gt;
**Bioconductor version**: `r BiocInstaller::biocVersion()`
&lt;br /&gt;
**Package version**: `r packageVersion("variants")`
&lt;/p&gt;


    </code>
</pre>
</div>
<div id="tidying-package-loading-output" class="section level3" number="24.3.3">
<h3>
<span class="header-section-number">24.3.3</span> Tidying package loading output<a class="anchor" aria-label="anchor" href="#tidying-package-loading-output"><i class="fas fa-link"></i></a>
</h3>
<p>Most workflows load a number of packages and you do not want
the output of loading those packages to clutter your workflow
document. Here’s how you would solve this in markdown; you can
do something similar in Latex.</p>
<p>First, set up a code chunk that is evaluated but not echoed, and whose
results are hidden. We also set <code>warning=FALSE</code> to be sure that
no output from this chunk ends up in the document:</p>
<pre><code>```{r, echo=FALSE, results="hide", warning=FALSE}
suppressPackageStartupMessages({
library(GenomicRanges)
library(GenomicAlignments)
library(Biostrings)
library(Rsamtools)
library(ShortRead)
library(BiocParallel)
library(rtracklayer)
library(VariantAnnotation)
library(AnnotationHub)
library(BSgenome.Hsapiens.UCSC.hg19)
library(RNAseqData.HNRNPC.bam.chr14)
})
```</code></pre>
<p>Then you can set up another code chunk that <em>is</em> echoed,
which has almost the same contents. The second invocation of <code><a href="https://rdrr.io/r/base/library.html">library()</a></code>
will not produce any output since the package has already been loaded:</p>
<pre><code>```{r}
library(GenomicRanges)
library(GenomicAlignments)
library(Biostrings)
library(Rsamtools)
library(ShortRead)
library(BiocParallel)
library(rtracklayer)
library(VariantAnnotation)
library(AnnotationHub)
library(BSgenome.Hsapiens.UCSC.hg19)
library(RNAseqData.HNRNPC.bam.chr14)
```</code></pre>
</div>
<div id="citations" class="section level3" number="24.3.4">
<h3>
<span class="header-section-number">24.3.4</span> Citations<a class="anchor" aria-label="anchor" href="#citations"><i class="fas fa-link"></i></a>
</h3>
<p>To manage citations in your workflow document,
specify the bibliography file in the document metadata header.</p>
<pre><code>bibliography: references.bib</code></pre>
<p>You can then use citation keys in the form of <code>[@label]</code> to cite an entry with an identifier “label.”</p>
<p>Normally, you will want to end your document with a section header “References” or similar, after which the bibliography will be appended.</p>
<p>For more details see the <a href="http://rmarkdown.rstudio.com/authoring_pandoc_markdown.html#citations">rmarkdown documentation</a>.</p>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="shiny.html"><span class="header-section-number">23</span> Shiny apps</a></div>
<div class="next"><a href="package-end-of-life-policy.html"><span class="header-section-number">25</span> Package End of Life Policy</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#non-software"><span class="header-section-number">24</span> Non-Software Packages</a></li>
<li><a class="nav-link" href="#annotation-and-experiment-data-packages"><span class="header-section-number">24.1</span> Annotation and Experiment data packages</a></li>
<li><a class="nav-link" href="#annotationexperiment-hub-packages"><span class="header-section-number">24.2</span> Annotation/Experiment Hub packages</a></li>
<li>
<a class="nav-link" href="#workflow-packages"><span class="header-section-number">24.3</span> Workflow packages</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#how-do-i-write-and-submit-a-workflow-vignette"><span class="header-section-number">24.3.1</span> How do I write and submit a workflow vignette?</a></li>
<li><a class="nav-link" href="#consistent-formatting"><span class="header-section-number">24.3.2</span> Consistent formatting</a></li>
<li><a class="nav-link" href="#tidying-package-loading-output"><span class="header-section-number">24.3.3</span> Tidying package loading output</a></li>
<li><a class="nav-link" href="#citations"><span class="header-section-number">24.3.4</span> Citations</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/Bioconductor/pkgrevdocs/blob/master/non-software-packages.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/Bioconductor/pkgrevdocs/edit/master/non-software-packages.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Bioconductor Packages: Development, Maintenance, and Peer Review</strong>" was written by Kevin Rue-Albrecht, Daniela Cassol, Johannes Rainer, Lori Shepherd. It was last built on 2022-01-31.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
